### DISCUSSION TITLE
What's the verification method when implementing microarchitectural change?

### DISCUSSION THREAD

**Discussion by nmd1411** [ChampSim/ChampSim#489] [General]
Hello,

I wanted to ask how I can verify change when integrating a design, for example memory renaming. I want to make sure that all retire instructions produce the correct value after I make such change. I couldn't find any compare method in the retire_rob stage in the code. 
Can you explain the verification concept of Champsim?

Thank you,

Nhat

**Comment by ngober** [comment]
Are you asking about verifying the data values following a load? Champsim does not model the data values in memory, it only models cache lines in the abstract.

**Reply by nmd1411** [reply]
I'm asking about any instruction that can generate value, change program control. Let's say if I want to predict value of an add instruction (value prediction) and I forget to do pipeline flush if it's a misprediction. Is there any check from Champsim to indicate that has happened? I would think that there should be some check at the retire stage for that.

**Reply by ngober** [reply]
Champsim also does not model values in registers, it only models the registers abstractly. I think the answer to your question is no, but only because the things you're asking about have up to this point been viewed as out-of-scope.

**Reply by nmd1411** [reply]
Thank you for the information.

**Reply by nmd1411** [reply]
I have another question: For memory disambiguation (store-to-load forwarding), does Champsim issue a memory request to cache or it just wait until the disambiguation is resolved if there's a store-load pair.

Thanks

**Reply by ngober** [reply]
No, because the memory address is stored in the trace, there is no ambiguity regarding store-load forwarding. ChampSim waits until the store data is executed, then forwards.

**Reply by nmd1411** [reply]
Your answer has clarified my misunderstanding. Thanks!
